taichi/platform/opengl/opengl_api.cpp

```diff
@@ -0,0 +1,270 @@
+    glShaderSource(id_, 1, &source_cstr, nullptr);
+
+    glCompileShader(id_);
+    GLint status = GL_TRUE;
+    glGetShaderiv(id_, GL_COMPILE_STATUS, &status);
+    if (status != GL_TRUE) {
+      GLsizei logLength;
+      glGetShaderiv(id_, GL_INFO_LOG_LENGTH, &logLength);
+      GLchar *log = new GLchar[logLength];
+      glGetShaderInfoLog(id_, logLength, &logLength, log);
+      TI_ERROR("[glsl] error while compiling shader:\n{}\n{}",
+          add_line_markers(source), log);
+    }
+    return *this;
+  }
+};
+
+struct GLProgram
+{
+  GLuint id_;
+
+  GLProgram()
+  {
+    id_ = glCreateProgram();
+  }
+
+  GLProgram(GLShader &shader)
+    : GLProgram()
+  {
+    this->attach(shader);
+  }
+
+  ~GLProgram()
+  {
+    glDeleteProgram(id_);
+  }
+
+  GLProgram &attach(GLShader &shader)
+  {
+    glAttachShader(id_, shader.id_);
+    return *this;
+  }
+
+  GLProgram &link()
+  {
+    glLinkProgram(id_);
+    GLint status = GL_TRUE;
+    glGetProgramiv(id_, GL_LINK_STATUS, &status);
+    if (status != GL_TRUE) {
+      GLsizei logLength;
+      glGetProgramiv(id_, GL_INFO_LOG_LENGTH, &logLength);
+      GLchar *log = new GLchar[logLength];
+      glGetProgramInfoLog(id_, logLength, &logLength, log);
+      TI_ERROR("[glsl] error while linking program:\n{}", log);
+    }
+    return *this;
+  }
+
+  GLProgram &use()
+  {
+    glUseProgram(id_);
+    return *this;
+  }
+
+  template <typename T>
+  void set_uniform(std::string name, T value)
+  {
+    GLuint loc = glGetUniformLocation(id_, name.c_str());
+    glapi_set_uniform(loc, value);
+  }
+};
+
+// https://blog.csdn.net/ylbs110/article/details/52074826
+// https://www.khronos.org/opengl/wiki/Shader_Storage_Buffer_Object
+// This is Shader Storage Buffer, we use it to share data between CPU & GPU
+struct GLSSBO
+{
+  GLuint id_;
+
+  GLSSBO()
+  {
+    glGenBuffers(1, &id_);
+  }
+
+  ~GLSSBO()
+  {
+    glDeleteBuffers(1, &id_);
+  }
+
+  /***
+   GL_{frequency}_{nature}:
+
+
+   STREAM
+       The data store contents will be modified once and used at most a few times.
+
+   STATIC
+       The data store contents will be modified once and used many times.
+
+   DYNAMIC
+       The data store contents will be modified repeatedly and used many times.
+
+
+   DRAW
+       The data store contents are modified by the application, and used as the source
+       for GL drawing and image specification commands.
+
+   READ
+       The data store contents are modified by reading data from the GL, and used to
+       return that data when queried by the application.
+
+   COPY
+       The data store contents are modified by reading data from the GL, and used as the
+       source for GL drawing and image specification commands.
+   ***/
+
+  GLSSBO &bind_data(void *data, size_t size, GLuint usage = GL_STATIC_READ)
+  {
+    glBindBuffer(GL_SHADER_STORAGE_BUFFER, id_);
+    glBufferData(GL_SHADER_STORAGE_BUFFER, size, data, usage);
+    return *this;
+  }
```

Comment:
I made some classes `GLSSBO, GLShader, GLProgram` to OOish GL's state machine API :)

Reply:
Yeah using OpenGL is really hard for beginners. Fortunately, with the OpenGL compute shader backend of Taichi, everyone who knows Python can play with OpenGL GPGPU!